File: src/main/java/com/jaadla/namlogapi/service/LogService.java
package com.jaadla.namlogapi.service;

import com.jaadla.namlogapi.model.Message;
import com.jaadla.namlogapi.repository.MessageRepository;
import lombok.extern.apachecommons.CommonsLog;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

@CommonsLog
@Service
public class LogService {

  MessageRepository messageRepository;

  public LogService(@Autowired MessageRepository messageRepository) {
    this.messageRepository = messageRepository;
  }

  public Page<Message> getMessages(String username, Pageable pageable) {

    return messageRepository.findByUsernameIgnoreCaseOrderByIdAsc(
        username, pageable);
  }
}

File: src/main/java/com/jaadla/namlogapi/LogService.kt
package com.jaadla.namlogapi

import com.jaadla.namlogapi.log.MessageRepository
import org.springframework.stereotype.Service

@Service
class LogService(
        private val messageRepository: MessageRepository
) {
    fun getMessages(username: String, startIndex: Int, size: Int): LogController.Page {
        var index = startIndex
        var totalMessages = -1
        if (index <= 0) {
            //Since the order is descending it would be impossible to know the last id when the first request is made.
            index = messageRepository.maxId() + 10
            //message count is only included in the first request as it's expensive.
            //for the "all" request it is sufficient to just use the max id for an approximate value.
            totalMessages = if (username.equals("all", ignoreCase = true)) index else messageRepository.getMessageCountUser(username)
        }
        val messages = if (username.equals("all", ignoreCase = true)) messageRepository.findAllMessages(index, size) else messageRepository.findByUsername(username, index, size)
        val finalPage = messages.size < size
        val lastId = if (messages.isNotEmpty()) messages[messages.size - 1].id!! else 0
        return LogController.Page(messages, lastId, finalPage, totalMessages)
    }
}
